apiserver/facades/client/modelgeneration/modelgeneration.go

```diff
@@ -127,23 +145,29 @@ func (m *ModelGenerationAPI) AdvanceGeneration(arg params.AdvanceGenerationArg)
 		case names.UnitTagKind:
 			results.Results[i].Error = common.ServerError(generation.AssignUnit(tag.Id()))
 		default:
-			results.Results[i].Error = common.ServerError(errors.Errorf("expected names.UnitTag or names.ApplicationTag, got %T", tag))
+			results.Results[i].Error = common.ServerError(
+				errors.Errorf("expected names.UnitTag or names.ApplicationTag, got %T", tag))
 		}
-		err = generation.Refresh()
-		if err != nil {
-			return results, errors.Trace(err)
+		if err := generation.Refresh(); err != nil {
+			return params.AdvanceGenerationResult{AdvanceResults: results}, errors.Trace(err)
 		}
 	}
 
+	result := params.AdvanceGenerationResult{AdvanceResults: results}
```

Comment:
Perhaps we should not proceed past here (checking for auto-completing) at all if there were errors above.

Reply:
Its an interesting question. What should happen if you "juju advance-generation exists noexists" ?
Should it advance 'exists' or should it fail with the state of the world being inconsistent with your request...

I would probably say that calling "advance" on something already on next probably shouldn't be an error (but probably a warning?) and should advance the other things in the list.

People may be likely to drive advance with a script, so what is good behavior there...

I don't feel like things will be changing underneath you (number of units, etc), such that you'll commonly be advancing something that has already died. So I feel like making advance feel more 'transactional' such that either it applies to everything or nothing probably feels a bit better. (with the caveat that advancing something already on next doesn't error, but does let you know it isn't going to do anything).

Thoughts?

Reply:
I guess this isn't actually transactional, because if you do "juju advance a b c" and c gets an error, then a and b have already moved forward, right?
So then there isn't much point to not continuing forward, as you're already in a partially applied state. If "c" can't be updated either it will block completing the generation, or it doesn't exist in which case we don't want a completed-but-not-removed generation, right?

Reply:
Advancing a unit already on the next generation is a no-op, so we will not get an error there.

Good point on the _not-exists_ scenario. I will leave the completion check in place, including where there were advance errors.